# Graph Data Science workshop with Neo4j

Click on the link below to open a Colab version of the notebook. You will be able to create your own version.

<table align="left">

  <td>
    <a href="https://colab.research.google.com/github/neo4j-field/graph-summit-apac-2023/blob/main/GDS_Workshop.ipynb" target="_blank">
      <img src="https://cloud.google.com/ml-engine/images/colab-logo-32px.png" alt="Colab logo">Run your own notebook in Colab
    </a>
  </td>
</table>

---
## Target

Do fraud analysis on a group of persons and transactions using graphs and data science.  

## Context

This notebook allows you load a dataset based on an updated version of [Paysim](https://www.sisu.io/posts/paysim/).  
PaySim is an approximation using an agent-based model and some anonymized, aggregate transactional data from a real mobile money network operator to create synthetic financial data sets academics and hackers can use for exploring ways to detect fraudulent behavior.  
Using this [code](https://github.com/voutilad/paysim), you can generate your own dataset with different caracteristics (size, fraud occurences...).    

We're going to leverage [Neo4j Graph Data Science (GDS)](https://neo4j.com/docs/graph-data-science/current/algorithms/) to investigate through the data and find out fraud patterns and fraudsters.  

## Dataset

The dataset used in this notebook represents money transafers between around 2500 clients, 75 merchants, 5 banks with 175000 transactions across 30 days.  
There are 5 types of transactions:  
* CashIn: a client moves money into the network via a merchant
* CashOut: a client moves money out of the network via a merchant
* Debit: a client moves money into a bank
* Transfer: a client sends money to another client
* Payment: a client exchanges money for something via a merchant

We will try to identify Clients which are fraudsters, trying to target other clients by taking their money below limits ot be unnoticed.  
We added from the the original Paysim some clients details (Phone, Email, SSN) to identify fake profiles too.  

---

## Let's get a graph database

We will use a Neo4j graph database created on the [Neo4j sandbox](https://neo4j.com/sandbox/).  
Once connected, on the _Select a project_ page, go to the section _Your own data_ and select the _Blank Sandbox_.  
Click on the _Create_ button at the bottom of the page.  
After few seconds, you should see the below.  
<img src="./img/sandbox_start.png" alt="Sandbox Start" width="50%" height="50%" title="Sandbox Start">  

And once it's up and running, you can access the connection details by clicking on the top right down arrow and picking the *Connection details* tab.  
You will need 2 things:
* Password  
* Bolt URL   

<img src="./img/sandbox_details.png" alt="Sandbox Details" width="50%" height="50%" title="Sandbox Details">  

---

## Let's code

First we will import the [Neo4j GDS python library](https://pypi.org/project/graphdatascience/)  

In [1]:
# Install Neo4j GDS Python Client
import sys
!{sys.executable} -m pip install graphdatascience

In [2]:
from graphdatascience import GraphDataScience

### Instantiate your GDS Session

Use Neo4j/Bolt URI and credentials according to your setup  

For local standalone instance Bolt connection without auth    
`gds = GraphDataScience("bolt://localhost:7687", auth=None)`  

For local standalone instance Bolt connection with auth    
`gds = GraphDataScience("bolt://localhost:7687", auth=("neo4j", "<password>"))`  

For remote cluster Neo4j connection with auth  
`gds = GraphDataScience("neo4j://<FQDN or IP Address>:7687", auth=("neo4j", "<password>"))`  

For remote standalone instance Bolt connection with auth   
`gds = GraphDataScience("bolt://<FQDN or IP Address>:7687", auth=("neo4j", "<password>"))` 

In [3]:
gds = GraphDataScience("bolt://3.231.156.13:7687", auth=("neo4j", "integrity-crusts-friction")) # >> Update the password and the URL here <<

### Check the GDS version installed

In [4]:
print(f"Neo4j GDS Version: {gds.version()}")

Neo4j GDS Version: 2.3.1


### Optional - Set database if you're not using the default _neo4j_ database. 

Not applicable for AuraDS, Neo4j Desktop and Neo4j Sandbox as we have only one database named _neo4j_.

In [5]:
#gds.set_database("my-db")

### Mark it to True if you want to erase all the data in the db and reload all.
If it is marked False, only the GDS properties will be remove, making the database ready for a rerun.

In [6]:
RELOAD_DATA = True

### Cleaning the database or making it ready for a rerun of the notebook.
Then we will use it to load the data from CSV files, running [Cypher](https://neo4j.com/developer/cypher/) queries.  

In [7]:
if RELOAD_DATA: # Delete all, takes few miniutes on a full database
    gds.run_cypher(
        """
        MATCH (n) CALL {
          WITH n
          DETACH DELETE n
        } IN TRANSACTIONS OF 10 ROWS;
        """
    )
else: # Reset the GDS properties when we re run the book without erasing all
    gds.run_cypher(
        """
        MATCH (c:Client) SET c.fraud_group = null, c.fraud_group_2 = null, c.score = null;
        """
    )

### Loading the data

Using [LOAD CSV](https://neo4j.com/docs/cypher-manual/current/clauses/load-csv/), we are loading csv files into the database, creationg the graph on the fly.  
The first cell is to test the file access, by reading it and showing only the first 5 rows.  

In [8]:
# Checking if we can access the data
if RELOAD_DATA:
    nodeListCSV = gds.run_cypher(
    """
    LOAD CSV WITH HEADERS FROM "https://raw.githubusercontent.com/neo4j-field/graph-summit-apac-2023/main/data/clients.csv" AS row
    RETURN row limit 5
    """
    )
    # The object returned is a Pandas Data Frame, so we can explore using standard Pandas methods
    # The output is not what we expected but remember here were reading the CSV file and loading it into the Data Frame
    nodeListCSV.head(5)

### Creating constraints

For data integrity, we will create [constraints](https://neo4j.com/docs/cypher-manual/current/constraints/) to have a robust graph data model.

In [9]:
if RELOAD_DATA:
    # First we create index
    CONSTRAINTS = [
      "CREATE CONSTRAINT ClientConstraint IF NOT EXISTS FOR (p:Client) REQUIRE p.id IS UNIQUE;",
      "CREATE CONSTRAINT MuleConstraint IF NOT EXISTS FOR (p:Mule) REQUIRE p.id IS UNIQUE;",
      "CREATE CONSTRAINT EmailConstraint IF NOT EXISTS FOR (p:Email) REQUIRE p.email IS UNIQUE;",
      "CREATE CONSTRAINT PhoneConstraint IF NOT EXISTS FOR (p:Phone) REQUIRE p.phoneNumber IS UNIQUE;",
      "CREATE CONSTRAINT SSNConstraint IF NOT EXISTS FOR (p:SSN) REQUIRE p.ssn IS UNIQUE;",
      "CREATE CONSTRAINT MerchantConstraint IF NOT EXISTS FOR (p:Merchant) REQUIRE p.id IS UNIQUE;",
      "CREATE CONSTRAINT BankConstraint IF NOT EXISTS FOR (p:Bank) REQUIRE p.id IS UNIQUE;",
      "CREATE CONSTRAINT TransactionConstraint IF NOT EXISTS FOR (p:Transaction) REQUIRE p.globalStep IS UNIQUE;",
      "CREATE CONSTRAINT DebitConstraint IF NOT EXISTS FOR (p:Transaction) REQUIRE p.globalStep IS UNIQUE;",
      "CREATE CONSTRAINT CashInConstraint IF NOT EXISTS FOR (p:CashIn) REQUIRE p.globalStep IS UNIQUE;",
      "CREATE CONSTRAINT CashOutConstraint IF NOT EXISTS FOR (p:CashOut) REQUIRE p.globalStep IS UNIQUE;",
      "CREATE CONSTRAINT TransferConstraint IF NOT EXISTS FOR (p:Transfer) REQUIRE p.globalStep IS UNIQUE;",
      "CREATE CONSTRAINT PaymentConstraint IF NOT EXISTS FOR (p:Payment) REQUIRE p.globalStep IS UNIQUE;"
    ]
    for c in CONSTRAINTS:
        gds.run_cypher(c)

### Loading all the data

We will load 7 csv files:  
* one for clients   
* one for merchants  
* five for transactions  

We can see how each node is created with a label and at leats one property.  
We see all the relationships between all the nodes, to show the money exchanges betweens all entities.  

In [10]:
if RELOAD_DATA:
    # Load Clients data
    gds.run_cypher(
    """
        LOAD CSV WITH HEADERS FROM "https://raw.githubusercontent.com/neo4j-field/graph-summit-apac-2023/main/data/clients.csv" AS row
        WITH row
        MERGE (c:Client { id: row.ID })
        SET c.name = row.NAME, c.isFraud = toBoolean(row.ISFRAUD)
        MERGE (p:Phone { phoneNumber: row.PHONENUMBER })
        MERGE (c)-[:HAS_PHONE]->(p)
        MERGE (s:SSN { ssn: row.SSN })
        MERGE (c)-[:HAS_SSN]->(s)
        MERGE (e:Email { email: row.EMAIL })
        MERGE (c)-[:HAS_EMAIL]->(e);
    """
    )

In [11]:
if RELOAD_DATA:
    # Load Merchants data
    gds.run_cypher(
    """
        LOAD CSV WITH HEADERS FROM "https://raw.githubusercontent.com/neo4j-field/graph-summit-apac-2023/main/data/merchants.csv" AS row
        WITH row
        MERGE (m:Merchant { id: row.ID })
        SET m.name = row.NAME, m.highRisk = toBoolean(row.HIGHRISK);
    """
    )

In [12]:
if RELOAD_DATA:
    # Load Debit data
    gds.run_cypher(
    """
        LOAD CSV WITH HEADERS FROM "https://raw.githubusercontent.com/neo4j-field/graph-summit-apac-2023/main/data/debit.csv" AS row
        WITH row
        MERGE (b:Bank { id: row.IDDEST })
        SET b.name = row.NAMEDEST
        MERGE (c:Client { id: row.IDORIG })
        MERGE (t:Transaction:Debit { globalStep: row.GLOBALSTEP })
        SET t.amount = toFloat(row.AMOUNT), t.isFraud = toBoolean(row.ISFRAUD)
        MERGE (t)-[:TO]->(b)
        MERGE (c)-[:PERFORMED]->(t);
    """
    )

In [13]:
if RELOAD_DATA:
    # Load CashIn data, largest file then takes few seconds
    gds.run_cypher(
    """
        LOAD CSV WITH HEADERS FROM "https://raw.githubusercontent.com/neo4j-field/graph-summit-apac-2023/main/data/cashin.csv" AS row
        CALL {
            WITH row
            MERGE (m:Merchant { id: row.IDDEST })
            SET m.name = row.NAMEDEST
            MERGE (c:Client { id: row.IDORIG })
            MERGE (t:Transaction:CashIn { globalStep: row.GLOBALSTEP })
            SET t.amount = toFloat(row.AMOUNT), t.isFraud = toBoolean(row.ISFRAUD)
            MERGE (t)-[:TO]->(m)
            MERGE (c)-[:PERFORMED]->(t)
        } IN TRANSACTIONS OF 10 ROWS;
    """
    )

In [14]:
if RELOAD_DATA:
    # Load CashOut data
    gds.run_cypher(
    """
        LOAD CSV WITH HEADERS FROM "https://raw.githubusercontent.com/neo4j-field/graph-summit-apac-2023/main/data/cashout.csv" AS row
        CALL {
            WITH row
            MERGE (m:Merchant { id: row.IDDEST })
            SET m.name = row.NAMEDEST
            MERGE (c:Client { id: row.IDORIG })
            SET c.name = row.NAMEORIG
            MERGE (t:Transaction:CashOut { globalStep: row.GLOBALSTEP })
            SET t.amount = toFloat(row.AMOUNT), t.isFraud = toBoolean(row.ISFRAUD)
            MERGE (t)-[:TO]->(m)
            MERGE (c)-[:PERFORMED]->(t)
        } IN TRANSACTIONS OF 10 ROWS;
    """
    )

In [15]:
if RELOAD_DATA:
    # Load Payment data
    gds.run_cypher(
    """
        LOAD CSV WITH HEADERS FROM "https://raw.githubusercontent.com/neo4j-field/graph-summit-apac-2023/main/data/payment.csv" AS row
        CALL {
            WITH row
            MERGE (m:Merchant { id: row.IDDEST })
            SET m.name = row.NAMEDEST
            MERGE (c:Client { id: row.IDORIG })
            SET c.name = row.NAMEORIG
            MERGE (t:Transaction:Payment { globalStep: row.GLOBALSTEP })
            SET t.amount = toFloat(row.AMOUNT), t.isFraud = toBoolean(row.ISFRAUD)
            MERGE (t)-[:TO]->(m)
            MERGE (c)-[:PERFORMED]->(t)
            FOREACH (ignoreMe in CASE WHEN row.TYPEORIG = 'MULE' THEN [1] ELSE [] END | SET c :Mule)
        } IN TRANSACTIONS OF 10 ROWS;
    """
    )

In [16]:
if RELOAD_DATA:
    # Load Transfer data
    gds.run_cypher(
    """
    LOAD CSV WITH HEADERS FROM "https://raw.githubusercontent.com/neo4j-field/graph-summit-apac-2023/main/data/transfer.csv" AS row
    CALL {
        WITH row
        MERGE (cd:Client { id: row.IDDEST })
        SET cd.name = row.NAMEDEST
        MERGE (co:Client { id: row.IDORIG })
        SET co.name = row.NAMEORIG
        MERGE (t:Transaction:Transfer { globalStep: row.GLOBALSTEP })
        SET t.amount = toFloat(row.AMOUNT), t.isFraud = toBoolean(row.ISFRAUD)
        MERGE (t)-[:TO]->(cd)
        MERGE (co)-[:PERFORMED]->(t)
        FOREACH (ignoreMe in CASE WHEN row.TYPEDEST = 'MULE' THEN [1] ELSE [] END | SET cd :Mule)
        FOREACH (ignoreMe in CASE WHEN row.TYPEORIG = 'MULE' THEN [1] ELSE [] END | SET co :Mule)
    } IN TRANSACTIONS OF 10 ROWS;
    """
    )

Before updating the graph, you can check out the result with this command in Neo4j Browser (accessible from the Sandbox UI).  
`CALL db.schema.visualization`  
to see the data model from the data you have just loaded

---
### Enriching the graph

Using the transaction details, we are able to enrich the model by adding the ordering of the transaction using the global step.

In [17]:
if RELOAD_DATA:
    # Update data model with new relationships
    gds.run_cypher(
    """
    MATCH (c:Client) with c.id as clientId
    CALL {
        WITH clientId
        MATCH (c:Client {id: clientId})-[:PERFORMED]->(tx:Transaction)
        WITH c, tx ORDER BY tx.globalStep
        WITH c, collect(tx) AS txs
        WITH c, txs, head(txs) AS _start, last(txs) AS _last

        MERGE (c)-[:FIRST_TX]->(_start)
        MERGE (c)-[:LAST_TX]->(_last)
        WITH c, apoc.coll.pairsMin(txs) AS pairs

        UNWIND pairs AS pair
          WITH pair[0] AS a, pair[1] AS b
          MERGE (a)-[n:NEXT]->(b)
    } IN TRANSACTIONS OF 10 ROWS;
    """
    )

### Having a first look at the dataset

In [18]:
result = gds.run_cypher(
    """
    CALL db.labels() YIELD label
    CALL apoc.cypher.run('MATCH (:`'+label+'`) RETURN count(*) as freq',{}) YIELD value
    WITH label,value.freq AS freq
    CALL apoc.meta.stats() YIELD nodeCount
    WITH *, 10^3 AS scaleFactor, toFloat(freq)/toFloat(nodeCount) AS relFreq
    RETURN label AS nodeLabel,
        freq AS frequency,
        round(relFreq*scaleFactor)/scaleFactor AS relativeFrequency
    ORDER BY freq DESC
    """
)
result

nodeLabel  frequency  relativeFrequency
0           Transaction     167943              0.954
1                CashIn      75231              0.427
2               Payment      34465              0.196
3               CashOut      33828              0.192
4              Transfer      22419              0.127
5                Client       2583              0.015
6                 Debit       2000              0.011
7                 Phone       1829              0.010
8                   SSN       1821              0.010
9                 Email       1814              0.010
10                 Mule       1583              0.009
11             Merchant         75              0.000
12                 Bank          4              0.000
13                  Mul          0              0.000
14  _Bloom_Perspective_          0              0.000

### Let's have a look on how the money is exchanged across entities

In [19]:
result = gds.run_cypher(
    """
    MATCH (t:Transaction)
    WITH sum(t.amount) AS globalSum, count(t) AS globalCnt
    WITH *, 10^3 AS scaleFactor
    UNWIND ['CashIn', 'CashOut', 'Payment', 'Debit', 'Transfer'] AS txType
        CALL apoc.cypher.run('MATCH (t:' + txType + ') RETURN sum(t.amount) as txAmount, count(t) AS txCnt', {}) YIELD value
        RETURN txType,
            value.txAmount AS TotalMarketValue,
            100 * round(scaleFactor * (toFloat(value.txAmount) / toFloat(globalSum)))/scaleFactor AS `%MarketValue`,
            100 * round(scaleFactor * (toFloat(value.txCnt) / toFloat(globalCnt)))/scaleFactor AS `%MarketTransactions`,
            toInteger(toFloat(value.txAmount) / toFloat(value.txCnt)) AS AvgTransactionValue,
            value.txCnt AS NumberOfTransactions
        ORDER BY `%MarketTransactions` DESC
    """
)
result

txType  TotalMarketValue  %MarketValue  %MarketTransactions  \
0    CashIn      1.044044e+10          45.5                 44.8   
1   Payment      2.904321e+08           1.3                 20.5   
2   CashOut      4.273458e+09          18.6                 20.1   
3  Transfer      7.950018e+09          34.6                 13.3   
4     Debit      1.168293e+07           0.1                  1.2   

   AvgTransactionValue  NumberOfTransactions  
0               138778                 75231  
1                 8426                 34465  
2               126329                 33828  
3               354610                 22419  
4                 5841                  2000

---
### Let's do some data science 

Now that our graph is constructed and filled with data, we can start our analysis.  

We will use the [Weakly Connected Components](https://neo4j.com/docs/graph-data-science/current/algorithms/wcc/) algorithm to investigate.  

---

In [20]:
# My first graph project name to use wcc algorithm
graphName = 'wccGroups'
# We will focus on fraud group above 5 members
fraudGroupMinSize = 4

In [21]:
# Remove existing projection with the same name, in case of a re run of the notebook
if gds.graph.exists(graphName).exists:
    gds.graph.drop(gds.graph.get(graphName))

### Lets' find first party fraud

By checking connections among clients, we can identify fake profiles and shady clients.  
We will start with a memory estimate of our projection.  

In [22]:
gds.graph.project.estimate(
    ['Client', 'SSN', 'Email', 'Phone'],     # Nodes to be added in the projection
    ['HAS_SSN', 'HAS_EMAIL', 'HAS_PHONE'])   # Relationships to be added in the projection

requiredMemory                                 [2727 KiB ... 4040 KiB]
treeView             graph projection: [2727 KiB ... 4040 KiB]\n|--...
mapView              {'components': [{'components': [{'memoryUsage'...
bytesMin                                                       2792856
bytesMax                                                       4137328
nodeCount                                                         8047
relationshipCount                                                15498
heapPercentageMin                                                  0.1
heapPercentageMax                                                  0.1
Name: 0, dtype: object

### We will create the projection to be used

We will use the variable _projection_ in our notebook. If you want to use it in Neo4j Browser, you can run 

In [23]:
projection, projectionPandas = gds.graph.project(graphName, ['Client', 'SSN', 'Email', 'Phone'], ['HAS_SSN', 'HAS_EMAIL', 'HAS_PHONE'])
projectionPandas

nodeProjection            {'Email': {'label': 'Email', 'properties': {}}...
relationshipProjection    {'HAS_SSN': {'orientation': 'NATURAL', 'indexI...
graphName                                                         wccGroups
nodeCount                                                              8047
relationshipCount                                                      7749
projectMillis                                                           369
Name: 0, dtype: object

### Before running the algorithm, we estimate the memory needed to execute it

In [24]:
gds.wcc.stream.estimate(projection)

requiredMemory                                                  62 KiB
treeView             Memory Estimation: 62 KiB\n|-- algorithm: 62 K...
mapView              {'components': [{'components': [{'components':...
bytesMin                                                         64472
bytesMax                                                         64472
nodeCount                                                         8047
relationshipCount                                                 7749
heapPercentageMin                                                  0.1
heapPercentageMax                                                  0.1
Name: 0, dtype: object

### Let's check how the algorithm will find community in the projection

In [25]:
result = gds.wcc.stream(projection)
result.head(10)

nodeId  componentId
0  176069            0
1  176070            1
2  176071            2
3       0            3
4       1            4
5       2            5
6       3            6
7       4            7
8       5            8
9       6            9

In [26]:
result.groupby(['componentId']).count().sort_values('nodeId', ascending=False).head(10)

nodeId
componentId        
5649             25
5207             25
5231             23
5191             23
5221             23
6164             22
5218             21
5698             21
5631             21
5626             21

### Let's populate the graph with a new property to represent the community detected by the wcc algorithm.

In [27]:
result_wcc = gds.wcc.write(projection, writeProperty='fraud_group', minComponentSize=fraudGroupMinSize)
result_wcc

writeMillis                                                            493
nodePropertiesWritten                                                 2631
componentCount                                                        1527
componentDistribution    {'p99': 20, 'min': 4, 'max': 25, 'mean': 5.269...
postProcessingMillis                                                   181
preProcessingMillis                                                      0
computeMillis                                                           16
configuration            {'jobId': '8bdea72e-a7d7-46aa-8ec2-924123eadab...
Name: 0, dtype: object

In [28]:
# Create an index on the new property just created by the wcc algorithm on Clients
gds.run_cypher("CREATE INDEX ClientFraudIndex IF NOT EXISTS FOR (c:Client) on c.fraud_group;")

Empty DataFrame
Columns: []
Index: []

In [29]:
# Look at the community created by the algorithm
# We can see the biggest community has 16 elements
result = gds.run_cypher("""
  MATCH (c:Client) WHERE c.fraud_group IS NOT NULL
  WITH c.fraud_group AS groupId, collect(c.id) AS members
  WITH groupId, size(members) AS groupSize
  WITH collect(groupId) AS groupsOfSize, groupSize
  RETURN groupSize, size(groupsOfSize) AS numOfGroups, groupsOfSize as FraudGroupIds
  ORDER BY groupSize DESC;
""")
result.head(10)

groupSize  numOfGroups                                      FraudGroupIds
0         16            2                                       [5207, 5649]
1         14            3                                 [5191, 5221, 5231]
2         13            1                                             [6164]
3         12            7         [5201, 5211, 5218, 5626, 5629, 5631, 5698]
4         11            3                                 [5230, 5239, 6152]
5         10           19  [5192, 5199, 5210, 5214, 5217, 5224, 5228, 524...
6          9           14  [5195, 5220, 5223, 5241, 5636, 5639, 5650, 569...
7          8           20  [5204, 5209, 5215, 5227, 5635, 5642, 5656, 567...
8          7           22  [5189, 5194, 5219, 5238, 5244, 5634, 5653, 567...
9          6           30  [5196, 5197, 5202, 5206, 5213, 5216, 5233, 524...

# Look at the large community in Neo4j Browser
Run this Cypher query in Neo4j Browser to visualize the two largest communities.  

`MATCH (c:Client)-[r:HAS_PHONE|HAS_EMAIL|HAS_PHONE]->(n) WHERE c.fraud_group IN [5207, 5649] RETURN c,r,n`  
or just to see large groups  
`MATCH (c:Client) WHERE c.fraud_group IS NOT NULL
WITH c.fraud_group AS groupId, collect(c.id) AS members
WITH *, size(members) AS groupSize WHERE groupSize > 12
MATCH p=(c:Client {fraud_group:groupId})-[:HAS_SSN|HAS_EMAIL|HAS_PHONE]->()
RETURN p`

In [30]:
result = gds.run_cypher("""
  MATCH (c:Client) WHERE c.fraud_group IS NOT NULL
  WITH c.fraud_group AS groupId, collect(c.id) AS members
  WITH groupId, size(members) AS groupSize WHERE groupSize > $gs
  MATCH (:Client {fraud_group:groupId})-[]-(txn)-[]-(d)      
  WHERE d.fraud_group IS NULL
  UNWIND labels(txn) AS txnType                                                 
  RETURN distinct(txnType), count(txnType);
""", params= {'gs': fraudGroupMinSize} )
result

txnType  count(txnType)
0  Transaction            4485
1     Transfer            4485

### Let's create a new property to identofy suspect clients

In [31]:
result = gds.run_cypher("""
  MATCH (c:Client) WHERE c.fraud_group IS NOT NULL
  WITH c.fraud_group AS groupId, collect(c.id) AS members
  WITH groupId, size(members) AS groupSize WHERE groupSize > $gs
  MATCH (c1:Client {fraud_group:groupId})-[]-(t:Transaction)-[]-(c2:Client)     
  WHERE c2.fraud_group IS NULL
  SET c1.suspect = true, c2.suspect = true                                      
  MERGE (c1)-[r:TRANSACTED_WITH]->(c2)                                          
  ON CREATE SET r += t
  RETURN count(r);
""", params= {'gs': fraudGroupMinSize})
result

count(r)
0      3535

### We found 3335 suspect transactions, let's investigate them using again wcc

In [32]:
graphName2 = 'wccGroups2'

In [33]:
# Remove existing graph with the same name
if gds.graph.exists(graphName2).exists:
    gds.graph.drop(gds.graph.get(graphName2))

### Creating a new projection using only suspect clients

In [34]:
projection2, projectionPandas2 = gds.graph.project.cypher(graphName2, 
          'MATCH (c:Client {suspect:true}) RETURN id(c) AS id', 
          'MATCH (c1:Client {suspect:true})-[r:TRANSACTED_WITH]->(c2:Client) RETURN id(c1) AS source, id(c2) as target')
projectionPandas2

nodeQuery            MATCH (c:Client {suspect:true}) RETURN id(c) A...
relationshipQuery    MATCH (c1:Client {suspect:true})-[r:TRANSACTED...
graphName                                                   wccGroups2
nodeCount                                                         1278
relationshipCount                                                 1165
projectMillis                                                       95
Name: 0, dtype: object

In [35]:
result = gds.wcc.write(projection2, writeProperty='fraud_group_2');

In [36]:
# Create an index on the new property
gds.run_cypher("CREATE INDEX ClientFraud2Index IF NOT EXISTS FOR (c:Client) on c.fraud_group_2;")

Empty DataFrame
Columns: []
Index: []

In [37]:
result = gds.run_cypher("""
MATCH (c:Client) WHERE c.fraud_group_2 IS NOT NULL
WITH c.fraud_group_2 AS secondGroupId, collect(c.id) AS members
RETURN secondGroupId, size(members) AS groupSize
ORDER BY groupSize DESC;
""")
result.head(5)

secondGroupId  groupSize
0             56         29
1             68         28
2              2         26
3             38         23
4             14         23

In [38]:
graphName3 = 'betweenness'

In [39]:
# Remove existing graph with the same name
if gds.graph.exists(graphName3).exists:
    gds.graph.drop(gds.graph.get(graphName3))

In [40]:
projection3, projectionPandas3 = gds.graph.project.cypher(graphName3, 
    'MATCH (c:Client) WHERE c.fraud_group_2 IS NOT NULL WITH c.fraud_group_2 AS secondGroupId, collect(c.id) AS members WITH secondGroupId, size(members) AS groupSize ORDER BY groupSize DESC LIMIT 1 WITH secondGroupId MATCH (c:Client {fraud_group_2:secondGroupId})-[r:TRANSACTED_WITH]-(c2:Client) RETURN id(c) AS id',
    'MATCH (c:Client) WHERE c.fraud_group_2 IS NOT NULL WITH c.fraud_group_2 AS secondGroupId, collect(c.id) AS members WITH secondGroupId, size(members) AS groupSize ORDER BY groupSize DESC LIMIT 1 WITH secondGroupId MATCH (c1:Client {fraud_group_2:secondGroupId})-[:TRANSACTED_WITH]-(c2:Client) RETURN id(c1) AS source, id(c2) AS target'
)

In [41]:
#Graph Projection Stats from Pandas object
print(projectionPandas3)

nodeQuery            MATCH (c:Client) WHERE c.fraud_group_2 IS NOT ...
relationshipQuery    MATCH (c:Client) WHERE c.fraud_group_2 IS NOT ...
graphName                                                  betweenness
nodeCount                                                           29
relationshipCount                                                   56
projectMillis                                                      386
Name: 0, dtype: object


In [42]:
result = gds.betweenness.write(projection3, writeProperty='score')
result

nodePropertiesWritten                                                    29
writeMillis                                                               8
centralityDistribution    {'p99': 520.003662109375, 'min': 0.0, 'max': 5...
postProcessingMillis                                                    797
preProcessingMillis                                                       0
computeMillis                                                            94
configuration             {'jobId': 'a2221c0d-e424-4a08-be50-ba72ba29e20...
Name: 0, dtype: object

In [43]:
gds.run_cypher("""
  CALL gds.betweenness.stream('betweenness') YIELD nodeId, score
  WITH gds.util.asNode(nodeId) AS c, score WHERE score > 0                       // Filter 0 scores again
  MATCH (c)-[r:TRANSACTED_WITH]-(:Client)                                        // Retrieve the relationships
  WITH c.name AS name, score, collect(r) AS txns                                 // Collect and count the number of relationships
  WITH name, score AS original, score/size(txns) AS newScore
  RETURN name, newScore, original ORDER BY newScore DESC;  
""")

name    newScore  original
0        Austin Cotton  173.333333     520.0
1        Kevin Bridges  171.000000     342.0
2         Tyler Turner  171.000000     342.0
3  Scarlett Stephenson  147.000000     294.0
4         John Cochran   46.750000     374.0
5     Dominic Hatfield   42.000000     294.0
6      Brooklyn Riddle   41.777778     376.0
7        Peyton Keller   27.000000      54.0